### PR TITLE
Blood filtration surgery removes toxins

### PR DESCRIPTION
**Pull request by xXPawnStarrXx**
## About The Pull Request

The basic premise of this is that bloodfiltering removes a small amount of toxin damage, to prevent this being abusable it will only repeat when they have reagents in their system. It works on slimes, however the trade off is that it's worse than medications (80% seiver at 1000k is like... 6 toxin per second, this is about 2.5 per second if my math is right) and will purge any meds in the subjects sytem. (Even the beneficial ones, use it wisely)

Currently; You _can_ heal toxins via surgery, however the stomach pump surgery cannot be used on many races, some of which are very commonly used by our players.

## Why It's Good For The Game
If you're filtering someone's blood, surely cycling clean blood around their system would purge some of the toxins (since that's kinda what our bloodstream does, in combination with our liver and kidneys). It is also convenient for those times where you are out of medications (Blackouts, breaches or it's otherwise unsafe to get them) and need to heal someone's toxins... click by click if required.

It's also much more humane than wrenching a carbon's stomach to make them sick and can work on slimes/other species that aren't listed as carbon.

## Proof Of Testing
image
image-1

## Changelog

:cl:
add: Blood filtering now restores toxins.
/:cl:

### CODE REVIEW

**Comment by xXPawnStarrXx** [comment]
Drafted for now as I try to work out why the hell it's not removing toxin/suffocation damage when it's literally reskinned code.

**Review by Majkl-J — CHANGES_REQUESTED** [review]
I see no reason for the oxygen one to exist, the methods to tend oxygen are everything from CPR, epinephrine (Medbay has like 200u lying around just roudstart alone), self resp, mediguns. Half of the mentioned already take absolutely no effort whatsoever.

As for toxin, sure, it tends to be a pain to fix sometime. Use `adjustToxLoss` for that, and proobably make it forced so doctors don't murder slime people?

**Comment by xXPawnStarrXx** [comment]
> I see no reason for the oxygen one to exist, the methods to tend oxygen are everything from CPR, epinephrine (Medbay has like 200u lying around just roudstart alone), self resp, other chems (Half are easy to make in a rush), mediguns. Half of the mentioned already take absolutely no effort whatsoever.

I more had the oxygen one in mind for people who can't breathe (Hemophages/synths) so they can't do CRP and as a holdover if someone's lungs had failed. I honestly don't see many doctors use much for suffocation besides CPR/Epi and if someone's allergic to oxy meds, just wanted the option there.

> As for toxin, sure, it tends to be a pain to fix sometime. Use `adjustToxLoss` for that, and proobably make it forced so doctors don't murder slime people?

I could try adjustToxLoss instead, I'll let you know.

**Comment by xXPawnStarrXx** [comment]
> As for toxin, sure, it tends to be a pain to fix sometime. Use `adjustToxLoss` for that, and proobably make it forced so doctors don't murder slime people?

Sadly, doesn't work. Not a blip in toxins either way

**Comment by Majkl-J** [comment]
> I more had the oxygen one in mind for people who can't breathe (Hemophages/synths) so they can't do CRP and as a holdover if someone's lungs had failed. I honestly don't see many doctors use much for suffocation besides CPR/Epi and if someone's allergic to oxy meds, just wanted the option there.

This was adressed on tg about a week ago with https://github.com/tgstation/tgstation/pull/81522

As for your other points, if your lungs fail the stasis beds prevent oxyloss, allowing you to stall for more than enough time to do a lobectomy. Epinephrine is blacklisted from allergies (And in fact is used to counter allergic reactions), so it is guaranteed to always work.

File: modular_zubbers/modules/surgery/healing.dm [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+	else
+		toxin_healed += round((target.getToxLoss() * toxin_multiplier),0.1)
+		oxygen_healed += round((target.getOxyLoss() * oxygen_multiplier),0.1)
+		dead_patient = FALSE
+	if(!get_location_accessible(target, target_zone))
+		toxin_healed *= 0.55
+		oxygen_healed *= 0.55
+		user_msg += " as best as you can while [target.p_they()] [target.p_have()] clothing on"
+		target_msg += " as best as [user.p_they()] can while [target.p_they()] [target.p_have()] clothing on"
+	target.heal_bodypart_damage(toxin_healed,oxygen_healed)
```
Comment (Majkl-J) [inline modular_zubbers/modules/surgery/healing.dm:96]:
```suggestion
	target.adjustToxLoss(toxin_healed, forced = TRUE)
	target.adjustOxyLoss(oxygen_healed)
```

**Review by Majkl-J — CHANGES_REQUESTED** [review]
heal_bodypart_damage only heals brute and burn

**Comment by xXPawnStarrXx** [comment]
> heal_bodypart_damage only heals brute and burn
I KNEW there was something I forgot.

**Comment by aKromatopzia** [comment]
Surgery for toxins already exist. Stomach pumping.

**Comment by Majkl-J** [comment]
> Surgery for toxins already exist. Stomach pumping

Oh yea that's true, I may have forgotten that one

https://github.com/Bubberstation/Bubberstation/blob/8e395113e855dc71e8ffa09b1d90818d682ae6c5/code/modules/mob/living/carbon/carbon.dm#L456

> why toxin damage is being repaired with a hemostat.

image

**Comment by xXPawnStarrXx** [comment]
> Surgery for toxins already exist. Stomach pumping. If you need one for corpses, just revive them on a stasis bed then pump their guts. Alternatively, you could make the blood filtering surgery do toxin healing, so nobody has to ask why toxin damage is being repaired with a hemostat.

I have it set to bloodfilter, also I don't want to remove reagents, I want one for toxin damage.

**Comment by xXPawnStarrXx** [comment]
For now, I've opened the bloodfiltering PR on TG itself, if it gets approved, I'll open a new PR to get it here quicker, if not. I will open it here as a //BUBBER EDIT//

**Comment by xXPawnStarrXx** [comment]
I have no idea why the maps are failing. I've changed so little now.

**Comment by nevimer** [comment]
don't worry about the maps, only meta and delta are needed to pass

**Comment by xXPawnStarrXx** [comment]
> don't worry about the maps, only meta and delta are needed to pass

I'm guessing screenshots comparison doesn't matter either, since... Y'know, they're not being changed.

**Comment by nevimer** [comment]
> > don't worry about the maps, only meta and delta are needed to pass
> 
> I'm guessing screenshots comparison doesn't matter either, since... Y'know, they're not being changed.

ill fix that for you

**Review by Majkl-J — APPROVED** [review]
Looks good for testmerging, might need to be lowered if it's too op but we can't tell without a TM.

**Comment by sneksnek** [comment]
The math on this PR is wrong. It takes a little around 2 seconds for a cycle to pass, resulting in one tick of metabolization. At 100% purity and 1000K temperature, seiver will heal 5 toxin each tick. This means it would be (give or take, depending on how the server is chugging), 2.5 toxin per second.

Completing a blood filtration takes 2.22 seconds on a stasis bed in default conditions (no upgrades, clean medbay, roundstart). At -2 toxin per operation, blood filtration would be 0.9 toxin/second.

Completing a stomach pump takes a flat 2 seconds on a stasis bed under the same conditions. It causes the patient to vomit, which heals 3 toxin. Effectively this is -3 toxin per operation, or 1.5 toxin/second.

A roundstart multiver pill + epinephrine heals 1.5 toxin/tick, or 0.75 toxin/second. Cryotubes heal about 2 toxin per tick at unupgraded temperatures, or 1 toxin/second.

I'm throwing these numbers out for comparison. As it is currently suggested by the PR with -2 toxin per operation, I think it is decently balanced. It's objectively worse than simply pumping the stomach, slightly worse than cryoxadone, moderately worse than seiver at maximal efficacy, and significantly worse than a proper IV mix. It fulfills the niche of healing toxin damage from a patient that lacks a stomach either due to organ loss or species specifications. One could argue that this will expediate the process of filtering someone's bloodstream whilst reducing their toxin damage as well now, but eh. They're already on a stasis bed likely and the next step was a stomach pump or a hit of seiver anyways. At any rate, it's still superior to purge with chemicals then heal with chemicals.

I deem this a SOUND PR and you have earned my THUMBS UP.

**Pull request merged**